cpp/src/rolling/rolling_detail.cuh

```diff
@@ -1333,14 +1070,20 @@ std::unique_ptr<column> rolling_window(column_view const& input,
   static_assert(warp_size == cudf::detail::size_in_bits<cudf::bitmask_type>(),
                 "bitmask_type size does not match CUDA warp size");
 
-  if (input.is_empty()) return empty_like(input);
+  if (input.is_empty()) { return empty_like(input); }
 
-  if (cudf::is_dictionary(input.type()))
+  if (cudf::is_dictionary(input.type())) {
```

Comment:
There's a lot of dictionary specific stuff scattered throughout here. Can't that be pushed into the preprocess/finalize steps?

Reply:
We could.  These checks are all nice and early though.

Reply:
The problem is that it's not consistent. Type specific logic should be handled in the pre/post-processing steps.

Reply:
After looking at this a bit:  This is a bit trickier than it seems.  The reason we're intercepting dictionary here and not later on is because we're fundamentally changing the type that gets passed down to int.

```
  auto input_col = cudf::is_dictionary(input.type())
                     ? dictionary_column_view(input).get_indices_annotated()
                     : input;
```

At a first glance, this seems like it might fit into the `preprocess`/`rolling`/`postprocess` abstraction, but it doesn't.  The preprocess step doesn't transmogrify input type A -> input type B.  It changes aggregation A -> aggregation B.   So there's no way to do this intercept without changing how the preprocessor thing works (which would entail changing groupby).

We _could_ Have the entire thing happen strictly in the postprocess step.   But this would be weird because then the postprocess step would essentially be reproducing most of the normal process itself (and on top of that, this is an aggregation-specific step so we'd have to either duplicate it per aggregation that allows dictionaries or make some shared functions.  Either way, it's a lot more code than what's here.

Reply:
We could potentially move the logic upwards into dispatch_rolling::operator().  Intercept input.type() == dictionary, and do the switch to integer at that point.  Then move the rest of the dictionary-specific logic into the postprocess step.

Reply:
Upon further inspection, I think it makes sense to defer any changes along these lines until later.  Potentially changing how the preprocess step works.